File: strings/zigzag-conversion/zigzag-conversion.java
class Solution {
  public String convert(String s, int numRows) {

    if (numRows == 1) {
      return s;
    }
    StringBuilder[] map = new StringBuilder[numRows];

    int i = 0;
    int j = 0;
    int n = s.length();
    int dir = -1;

    while (j < n) {
      if (map[i] == null) {
        map[i] = new StringBuilder();
      }
      StringBuilder builder = map[i];

      builder.append(s.charAt(j));

      if (i == numRows - 1 || i == 0) {
        dir = -dir;
      }

      i += dir;

      j++;
    }

    StringBuilder main = new StringBuilder();

    for (StringBuilder b : map) {
      if (b != null) {
        main.append(b.toString());
      }
    }

    return main.toString();

  }
}

File: strings/zigzag-conversion/zigzag-conversion.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
function convert(s, numRows) {
  if (numRows === 1) {
    return s;
  }

  const sbArr = new Array(numRows).fill(null).map(() => []);

  let i = 0;
  let j = 0;
  let n = s.length;
  let dir = -1;

  while (j < n) {
    const builder = sbArr[i];

    builder.push(s[j]);

    if (i === numRows - 1 || i === 0) {
      dir = -dir;
    }

    i += dir;

    j++;
  }

  const main = [];

  for (let i = 0; i < sbArr.length; i++) {
    main.push(sbArr[i].join(''));
  }

  return main.join('');
}
